components/security/audit/backend/src/main/java/org/phenotips/security/audit/internal/HibernateAuditStore.java

```diff
@@ -104,13 +118,151 @@ public void store(AuditEvent event)
+
+            return foundEntries;
+        } catch (HibernateException ex) {
+            this.logger.error("Failed to load audit event documents: {}", ex.getMessage(), ex);
+        }
+        return Collections.emptyList();
+    }
+
+    @Override
+    public long countEventsForEntity(DocumentReference entity)
+    {
+        return getCount(null, null, null, entity);
+    }
+
+    @Override
+    public long countEventsForEntity(DocumentReference entity, String type)
+    {
+        return getCount(null, null, type, entity);
+    }
+
+    @Override
+    public long countEventsForUser(User user)
+    {
+        return getCount(user, null, null, null);
+    }
+
+    @Override
+    public long countEventsForUser(User user, String ip)
+    {
+        return getCount(user, ip, null, null);
+    }
+
+    @Override
+    public long countEventsForUser(User user, String ip, String type)
+    {
+        return getCount(user, ip, type, null);
+    }
+
+    @Override
+    public long countEvents(AuditEvent eventTemplate, Calendar fromTime, Calendar toTime)
+    {
+        try {
+            Session session = this.sessionFactory.getSessionFactory().openSession();
+            Criteria c = session.createCriteria(AuditEvent.class);
+
+            if (eventTemplate != null) {
+                c.add(Example.create(eventTemplate));
+            }
+
+            setTimeInterval(c, fromTime, toTime);
+            c.setProjection(Projections.rowCount());
+
+            @SuppressWarnings("rawtypes")
+            List foundEntries = c.list();
+            if (foundEntries != null && !foundEntries.isEmpty()) {
+                long rowCount = (long) foundEntries.get(0);
+                return rowCount;
+            }
+
+            return 0;
+        } catch (HibernateException ex) {
+            this.logger.error("Failed to count audit event documents: {}", ex.getMessage(), ex);
+            return -1;
+        }
+    }
+
+    private long getCount(User user, String ip, String type, DocumentReference entity)
+    {
+        try {
+            Session session = this.sessionFactory.getSessionFactory().openSession();
+            Criteria c = session.createCriteria(AuditEvent.class);
+            AuditEvent sample = new AuditEvent(user, ip, type, null, entity, null);
+            c.add(Example.create(sample));
+            c.addOrder(Order.desc(TIME_FIELD_NAME));
+            c.setProjection(Projections.rowCount());
+
+            @SuppressWarnings("rawtypes")
+            List foundEntries = c.list();
+            if (foundEntries != null && !foundEntries.isEmpty()) {
+                long rowCount = (long) foundEntries.get(0);
+                return rowCount;
+            }
+
+            return 0;
+        } catch (HibernateException ex) {
+            this.logger.error("Failed to count audit event documents: {}", ex.getMessage(), ex);
+            return -1;
+        }
+    }
+
+    private void setTimeInterval(Criteria c, Calendar fromTime, Calendar toTime)
```

Comment:
It would be better to not enforce time between [Epoch .. NOW], and allow `from` and `to` to be null, in which case instead of adding a `between` filter, we'd add, as needed, one of `between`, `ge`, `le`, or not restriction at all.